File: leetcode/data_struct/string/0067/1.ts
// auther: phoenix
// date: 2023-06-15 14:49:49

/*

11
01

1010
1011



*/

// 方法一、模拟手工计算
addBinary1test()
function addBinary1test() {
  console.log(addBinary1('11', '1'))      // 100
  console.log(addBinary1('1010', '1011')) // 10101
}

function addBinary1 (a: string, b: string) : string {
  let sum:string = ''
  let add = 0 // 进位
  let maxLength = Math.max(a.length, b.length)
  a = a.padStart(maxLength, '0')
  b = b.padStart(maxLength, '0')

  for (let i = maxLength - 1; i >= 0 ; i--) {
    let sum1 = Number(a[i]) + Number(b[i]) + add
    switch (sum1) {
      case 0:
        sum += '0'
        add = 0
        break;
      case 1:
        sum += '1'
        add = 0
        break;
      case 2:
        sum += '0'
        add = 1
        break;
      case 3:
        sum += '1'
        add = 1
        break;
      default:
        break;
    }
  }
  sum = sum.split('').reverse().join('')
  if (add === 1) {
    sum = '1' + sum
  }

  return sum
}

// 方法二、进制转换
// addBinary2test()
function addBinary2test() {
  console.log(addBinary2('11', '1'))      // 100
  console.log(addBinary2('1010', '1011')) // 10101
}
function addBinary2 (a: string, b: string) : string {
  let a10 = parseInt(a, 2)
  let b10 = parseInt(b, 2)

  return (a10 + b10).toString(2) + ''
}

File: leetcode/data_struct/string/0067/1.rs
// auther: phoenix
// date: 2023-06-15 14:49:46

use std::cmp;
/*


    let s2 = format!("{1}是个体重{0:>0width$}KG，身高{height:?}cm的矮胖子",
                     115,
                     "肥仔",
                     width = 5,
                     height = 163);

肥仔是个体重00115KG，身高163cm的矮胖子

*/

fn main() {
  add_binary1_test();
}

fn add_binary1_test () {
  println!("{}", add_binary1(String::from("11"), String::from("1")));        // 100
  println!("{}", add_binary1(String::from("1010"), String::from("1011")));   // 10101
}

fn add_binary1(a: String, b:String) -> String {
  let mut sum:String = String::from("");
  let mut add:usize = 0;  // 进位
  let max_length:usize = cmp::max(a.len(), b.len());

  let mut s1:String = format!("{0:>0width$}", a, width = max_length);
  let mut s2:String = format!("{0:>0width$}", b, width = max_length);

  for i in (0..max_length).rev() {
    println!("-{} ", s1.get(i..i+1));
    // let sum1:usize = s1[i].to_



  }

  return sum
}


// https://leetcode.cn/problems/add-binary/solution/rust-yuan-li-wo-du-dong-yong-apiyi-xing-dai-ma-ke-/
// 方法二、进制转换
pub fn add_binary2(a: String, b: String) -> String {
  format!("{:b}", i128::from_str_radix(a.as_str(),2).unwrap() + i128::from_str_radix(b.as_str(),2).unwrap())
}
